src/views/NavigationEvents.js

```diff
@@ -0,0 +1,61 @@
+import React from 'react';
+import withNavigation from './withNavigation';
+
+const EventNameToPropName = {
+  willFocus: 'onWillFocus',
+  didFocus: 'onDidFocus',
+  willBlur: 'onWillBlur',
+  didBlur: 'onDidBlur',
+};
+
+const EventNames = Object.keys(EventNameToPropName);
+
+class NavigationEvents extends React.Component {
+  constructor() {
+    super();
+  }
+
+  componentDidMount() {
+    this.subscriptions = {};
+    EventNames.forEach(this.addListener);
+  }
+
+  componentDidUpdate(prevProps) {
+    EventNames.forEach(eventName => {
+      const listenerHasChanged =
+        this.props[EventNameToPropName[eventName]] !==
+        prevProps[EventNameToPropName[eventName]];
+      if (listenerHasChanged) {
+        this.removeListener(eventName);
+        this.addListener(eventName);
+      }
+    });
+  }
+
+  componentWillUnmount() {
+    EventNames.forEach(this.removeListener);
+  }
+
+  addListener = eventName => {
```

Comment:
as an alternative, we could consider

```
addListener = eventName => {
  this[eventName] = (...args) => {
    const listener = this.props[EventNameToPropName[eventName]];
    listener && listener(...args)
  }

  this.subscriptions[eventName] = this.props.navigation.addListener(
    eventName,
    this[eventName]
  );
};
```

Or, more verbosely (to show what I mean):

```
class NavigationEvents extends React.Component {
  willBlur = (...args) => {
    const { onWillBlur } = this.props;
    onWillBlur && onWillBlur(...args);
  };

  addListenerForWillBlur = () => {
    this.subscriptions.willBlur = this.props.navigation.addListener('willBlur', this.willBlur);
  };
}
```

This would allow to remove `componentDidUpdate` and the remove + add listener calls. Just subscribing once in `componentDidMount`, unsubscribing in `componentWillUnmount` and not caring about the changes of the props of `NavigationEvents`.

Reply:
@vonovak I've proposed a similar implementation initially, using "internal listeners", but @satya164 suggested I subscribe user-provided listeners directly [here](https://github.com/react-navigation/rfcs/issues/41#issuecomment-387016663)

I'm ok for your proposal, the only drawback is that if user provides a single callback, you would have 4 listeners being subscribed, with 3 of them doing nothing. Not a big deal IMHO so if we settle on that I'll make the change.

Reply:
Hmm, I think @satya164's suggestion makes a lot of sense and this implementation is quite good. It may be a tiny bit harder to read but it prevents creation of unnecessary upstream subscriptions